### PR TITLE
mtmd: support Qwen3-TTS (note: breaking change to llama-tts binary)

### PR DESCRIPTION
**Pull request by ngxson**
## Overview

Target support https://huggingface.co/Qwen/Qwen3-TTS-12Hz-1.7B-Base

## Qwen3-TTS

Available params:
- `--tts-lang` can be `zh`, `en`, `de`, `it`, `pt`, `es`, `ja`, `ko`, `fr`, `ru` (default: `en`)
- `--tts-speaker-file` should point to a speaker reference audio file (wav, mp3)

Example usage:

```sh
llama-tts -hf ggml-org/Qwen3-TTS-12Hz-1.7B-Base-GGUF \
    -p "Hello world" \
    --tts-lang english \
    --tts-speaker-file speaker.mp3 \
    --output out.wav
```

## API design choices

This Qwen3-TTS works mostly the same as Sesame CSM, see PR https://github.com/ggml-org/llama.cpp/pull/12648 :
- Audio reference is encoded to text embd space via an audio encoder (`speaker_encoder` for qwen; mimi encoder for sesame)
- Audio reference and prompt are decoded via a causal "backbone" model (`talker.model`)
- Backbone model sample the first codebook entry (aka semantic audio code) and expose the output embedding
- "Code predictor" model take the sampled semantic code & backbone embedding and generate the next 15 acoustic tokens (causal, similar to MTP model)
- Sum 16 codebook entries, then decode the summed value via backbone to generate the next audio code

These design choices are made to adapt this model to existing llama.cpp infrastructure:
- For `talker.model`:
    - `codec_embedding` is concat to the text embedding table, vocab is extended, example:
        - codec_bos_id(2149) --> "<|codec_bos|>"
        - codec_eos_token_id(2150) --> "<|codec_eos_token|>"
        - codec_language_id.chinese(2055) --> "<|codec_language_chinese|>"
        - other rows --> "<|codec_0|>", "<|codec_1|>", ..., "<|codec_1023|>"
    - output tensor `codec_head` is smaller than vocab, so logits will be padded at inference time
    - `suppress_tokens` is used to limit the backbone to only sample either semantic or EOS (stop) token
- For `speaker_encoder`: treat it as a normal audio encoder, using existing mtmd_audio infrastructure
- For `code_predictor`:
    - I don't reuse libllama or MTP infra for it, because (1) it will introduce quite a lot of hacky patches and (2) libllama doesn't support generating N tokens in-graph; it still needs to sync at every token which makes it quite slow (memory bound) for very small model like `code_predictor` (ref: https://github.com/ggml-org/llama.cpp/pull/12648#issuecomment-2768989097)
    - The model runs 15 steps + sampling to generate the 15 acoustic codes; everything is done on one single graph (one forward output 15 codes)
    - To make that possible, I implemented (1) backend sampling for top_k/p and (2) minimal on-graph kv management; kv is alloc once per forward call (we don't reuse it afterward anyway)

## Develooment

TODO - easy items (can reuse the existing infra):
- [x] convert talker backbone
- [x] convert speaker encoder
- [x] load speaker encoder via mtmd and impl the encoder pass

TODO - complex items (need to write new components):
- [x] handle `codec_embeddings` and codec control tokens --> append them to the vocab, see conversion script
- [x] handle `codec_head` --> use it as output head for backbone, pad logits output to match vocab; we also masked away codec control tokens from sampling via `suppress_tokens`
- [x] convert & pack the rest into mmproj
- [x] create `mtmd_gen` interface
- [x] impl `code_predictor`
- [x] impl `code2wav` in mtmd
- [x] add helper
- [x] revamp `llama-tts`
- [x] update dev docs

Follow-up PRs:
- add support to `llama-server`
- add warmup pass, make sure --fit handles it correctly

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: About 5-10% of core code (API design and some boilerplates) are human-coded, the rest is AI-generated to fit the target design

### CODE REVIEW

File: src/llama-ext.h [hunk 1]
```diff
@@ -124,3 +124,8 @@ LLAMA_API llama_context * llama_get_ctx_other(struct llama_context * ctx);
 LLAMA_API const int32_t * llama_model_target_layer_ids  (const struct llama_model * model);
 // returns the number of extracted layers from target model
 LLAMA_API uint32_t        llama_model_target_layer_ids_n(const struct llama_model * model);
+
+// retrieves the whole token embedding matrix in F32 format (n_embd x n_vocab)
+// returns total number of elements (usually n_embd * n_vocab) or 0 on error
+// if out is nullptr, returns the number of tokens without writing to out
+LLAMA_API uint32_t llama_model_get_tok_embd(const struct llama_model * model, float * out);
```
Comment (ngxson) [inline src/llama-ext.h:132]: @ggerganov note that I need to add this API because some models requires combining input token with an embedding (element-wise sum), so we still need this API even after the `llama_batch_ext` refactoring

File: tools/mtmd/mtmd.h [hunk 2]
```diff
@@ -327,6 +327,59 @@ struct mtmd_caps {
...
+    const int32_t * codes;
+    size_t n_codes;
+    const float * embd; // the generated hidden state, to be fed back to backbone
+                        // it must have n_text_embd elements
+
+    // for MTMD_GEN_PROCESS_TYPE_CODE2WAV
+    const float * audio;
+    size_t        n_samples;
+    const char * state_data;
+    size_t       state_size;
```
Comment (ngxson) [inline tools/mtmd/mtmd.h:377]: @ggerganov note about design choice here: it might be better to run causal models in `libllama` and use the proper kv cache impl there (similar to `wavtokenizer-dec`), however, I decide to do it in `clip.cpp` because:
- I don't want to introduce too many changes / hacks to `libllama` in this early stage
- kv cache of these models are usually very small, the perf impact is not significant compared to fully backend kv cache
- we still have an escape hatch for future: when `libllama` is ready for that, we can create an in-memory `gguf_context` and load it via `llama_model_init_from_user`

the current `clip.cpp` API design exports kv cache state after each run as raw bytes `state_data`, then next call set it as input

File: src/llama-ext.h (hunk 1 above)
Comment (ggerganov) [reply src/llama-ext.h:132]: Not sure if suitable in this case, but might be possible to utilize the `ctx_other`? Anyway, it is OK as proposed.

**Review by ggerganov — APPROVED** [review]
The `libllama` changes are good.

Consider adding the model to https://github.com/ggml-org/convert

**Comment by itn3rd77** [comment]
@ngxson Is the missing `--tts-lang` de (German) intentional and not supported?

File: src/llama-ext.h (hunk 1 above)
Comment (ngxson) [reply src/llama-ext.h:132]: it's not possible since the caller is `mtmd`, not another `llama_context`, so some sorts of new API would still be required

**Comment by ngxson** [comment]
@itn3rd77 I updated the docs to reflect the ISO 639-1 code, PTAL

File: tools/tts/tts.cpp [hunk 3]
```diff
@@ -1,1096 +1,194 @@
...
-
-    if (llama_encode(ctx_cts, batch) != 0) {
-        LOG_ERR("%s: llama_encode() failed\n", __func__);
+    if (gen.set_input(&inp) != 0) {
+        LOG_ERR("set_input failed\n");
         return 1;
     }
 
-    llama_synchronize(ctx_cts);
+    // TODO: explicit prompt processing step here
```
Comment (ngxson) [inline tools/tts/tts.cpp]: 2 things left before this PR can be merged (mostly to make it server-compatible, will come in follow-up PR):
- [x] have a dedicated `step()`-like func for prompt processing, so that server can interrupt if task is cancelled
- [x] add a gguf metadata to indicate that this model cannot be used for chat or text generation

File: conversion/qwen3tts.py [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+        ):
+            return None
+
+        return super().filter_tensors((name, gen))
+
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        # code2wav tensors come from generate_extra_tensors() already renamed to their
+        # final gguf name (bypassing tensor_mapping.py, same as the code_predictor
+        # tensors below); prepare_tensors() re-runs modify_tensors() on them too, so
+        # pass them through untouched instead of falling into map_tensor_name()
```
Comment (CISC) [inline conversion/qwen3tts.py]: ..if only there was some kind of rule against comment blocks like this... :D

**Review by CISC — APPROVED** [review]
The conversion code is messy, but acceptable for now.

Hopefully can be cleaned up/unified as we get more TTS models?

File: conversion/qwen3tts.py (hunk 4 above)
Comment (ngxson) [reply conversion/qwen3tts.py]: yes there is a set of rules (with examples) in AGENTS.md, but sadly they often get lost between context compaction:

> - Code comments:
>     - Keep code comments concise (usually 1-2 lines)
>     - Avoid redundant or excessive inline commentary
>     - Avoid hard-wrapping it to a fixed column width - that hurts readability
>     - Use ASD-STE100 Simplified Technical English, simple wordings (write like cavemen if needed)
>     - Note: Remind yourself of this point regularly, as it often gets lost between context compactions

**Comment by ngxson** [comment]
> Hopefully can be cleaned up/unified as we get more TTS models?

yes I've already planned to support chatterbox (@ServeurpersoCom is working on that) and pocket-tts. hopefully when we have more model, we can start to see common patterns and clean up / dedup some code

**Comment by ggerganov** [comment]
How long does the reference audio need to be usually?

File: conversion/qwen3tts.py (hunk 4 above)
Comment (CISC) [reply conversion/qwen3tts.py]: Yeah, I know, I was being snarky. :P

**Comment by ngxson** [comment]
> How long does the reference audio need to be usually?

usually that's mentioned on the model card, it seems to be 3s:

image

File: conversion/qwen3tts.py (hunk 4 above)
Comment (ngxson) [reply conversion/qwen3tts.py]: AI clean up AI comments: https://github.com/ggml-org/llama.cpp/pull/26254/commits/201e50cc2076a20adc460c41598593c7cd7b0813

File: conversion/qwen3tts.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+
+    def generate_extra_tensors(self) -> Iterable[tuple[str, Tensor]]:
+        yield from self._generate_code2wav_tensors()
+
+    def _generate_code2wav_tensors(self) -> Iterable[tuple[str, Tensor]]:
+        # code2wav weights live in speech_tokenizer/model.safetensors, not the main safetensors
+        from safetensors.torch import load_file
+
+        wav_config = self._wav_decoder_config()
+        state_dict = load_file(self.dir_model / "speech_tokenizer" / "model.safetensors")
```
Comment (ngxson) [inline conversion/qwen3tts.py:357]: @CISC this pattern is popping up recently as some models stores audio-related tensor in a subdir (this model and xiaomi's mimo are 2 examples). if you are interested, feel free to push a refactoring to generalize this logic

Comment (CISC) [reply conversion/qwen3tts.py:357]: Will look into it after I have worked through the backlog.

**Comment by woheller69** [comment]
Generates frames and then fails...

ggml-cpu/ops.cpp:4886: GGML_ASSERT(i01 >= 0 && i01 < ne01) failed

llama.cpp b10276 and Q8 files from here: https://huggingface.co/ggml-org/Qwen3-TTS-12Hz-1.7B-Base-GGUF/tree/main

**Comment by ServeurpersoCom** [comment]
> Generates frames and then fails...
> 
> ggml-cpu/ops.cpp:4886: GGML_ASSERT(i01 >= 0 && i01 < ne01) failed
> 
> llama.cpp b10276 and Q8 files from here: https://huggingface.co/ggml-org/Qwen3-TTS-12Hz-1.7B-Base-GGUF/tree/main

Please test this : https://github.com/ggml-org/llama.cpp/compare/master...ServeurpersoCom:llama.cpp:ggml/build-forward-order

It's the tested fix for CPU

**Comment by woheller69** [comment]
yes, that fixes it :-)

**Comment by quine00** [comment]
> usually that's mentioned on the model card, it seems to be 3s:

I'm not sure what to expect, but I tried cloning my own voice with reference audio, using invocations like,

```
./build/bin/llama-tts -hf ggml-org/Qwen3-TTS-12Hz-1.7B-Base-GGUF \
  -p "the quick brown fox jumped over the lazy dog" \
  --tts-lang english --tts-speaker-file /tmp/out.wav --output out.wav
```

Where `/tmp/out.wav` contained various reference texts spoken by myself ranging from 3s to 30s in length. In all cases, the resulting speech was no where near my original voice, which is a fairly standard British English accent. Is this a known limitation? The model card / description makes this feature sound like it can do a better job than I've managed to reproduce.

**Comment by Rye426** [comment]
Requesting support for Qwen3-TTS-12HZ-0.6B. While the code successfully exports the 0.6B model, an error occurs during inference:0.03.232.837 E clip_init: failed to load model 'mmproj-Qwen3-TTS-12Hz-0.6B-Base-F16.gguf': operator(): unable to find tensor a.gen.code.proj_in.weight

**Comment by ServeurpersoCom** [comment]
Hello,

@Rye426, this should be fixed here: https://github.com/ggml-org/llama.cpp/pull/28231. Note that the mmproj GGUF has to be regenerated, the F32 rule lives in the conversion script.

@quine00 the fix touches the same code path, so it is worth retrying once it lands. That said, cloning here is speaker-embedding only: we extract a single timbre vector from your reference clip and condition on that.

The reference implementation also has an in-context mode that feeds the clip's own audio codes plus its transcript, which is what gets you a close likeness, and that one needs the codec encoder, which the mmproj does not currently ship.

**Comment by Rye426** [comment]
> Hello,
> 
> @Rye426, this should be fixed here: #28231. Note that the mmproj GGUF has to be regenerated, the F32 rule lives in the conversion script.
> 
> @quine00 the fix touches the same code path, so it is worth retrying once it lands. That said, cloning here is speaker-embedding only: we extract a single timbre vector from your reference clip and condition on that.
> 
> The reference implementation also has an in-context mode that feeds the clip's own audio codes plus its transcript, which is what gets you a close likeness, and that one needs the codec encoder, which the mmproj does not currently ship.

Thank you!
But
There is another issue: when using the Vulkan backend to execute the `GET_ROWS` operator, the backend does not support the current tensor offset configuration.
llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:12111: GGML_ASSERT(dst->op != GGML_OP_GET_ROWS || (a_offset == 0 && b_offset == 0 && d_offset == 0)) failed
[New LWP 761440]
[New LWP 761439]
[New LWP 761438]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
0x0000ffffa16eef68 in ?? () from /lib/aarch64-linux-gnu/libc.so.6
#0  0x0000ffffa16eef68 in ?? () from /lib/aarch64-linux-gnu/libc.so.6
#1  0x0000ffffa16e2618 in ?? () from /lib/aarch64-linux-gnu/libc.so.6
#2  0x0000ffffa16e265c [PAC] in ?? () from /lib/aarch64-linux-gnu/libc.so.6
#3  0x0000ffffa173aa24 [PAC] in wait4 () from /lib/aarch64-linux-gnu/libc.so.6
#4  0x0000aaaabf25216c [PAC] in ggml_print_backtrace ()
#5  0x0000aaaabf2522b0 in ggml_abort ()
#6  0x0000aaaabf07af00 in void init_pushconst_tensor_offsets<vk_op_binary_push_constants>(ggml_backend_vk_context*, vk_op_binary_push_constants&, ggml_tensor const*, ggml_tensor const*, ggml_tensor const*, ggml_tensor const*, ggml_tensor*) ()
#7 0x0000aaaabf20101c in void ggml_vk_op_f32<vk_op_binary_push_constants>(ggml_backend_vk_context*, std::shared_ptr<vk_context_struct>&, ggml_tensor const*, ggml_tensor const*, ggml_tensor const*, ggml_tensor const*, ggml_tensor*, ggml_op, vk_op_binary_push_constants&&) [clone .constprop.0] ()
#8 0x0000aaaabf202998 in ggml_vk_get_rows(ggml_backend_vk_context*, std::shared_ptr<vk_context_struct>&, ggml_tensor const*, ggml_tensor const*, ggml_tensor*) ()
#9 0x0000aaaabf21c230 in ggml_vk_build_graph(ggml_backend_vk_context*, ggml_cgraph*, int, ggml_tensor*, int, bool, bool, bool) [clone .isra.0] ()
#10 0x0000aaaabf21d594 in ggml_backend_vk_graph_compute(ggml_backend*, ggml_cgraph*) ()
#11 0x0000aaaabf273d58 in ggml_backend_sched_compute_splits(ggml_backend_sched*) ()
#12 0x0000aaaabf27490c in ggml_backend_sched_graph_compute ()
#13 0x0000aaaabef21254 in clip_encode(clip_ctx*, clip_encode_params*) ()
#14 0x0000aaaabee8c354 in mtmd_gen_audio_process_impl(mtmd_context*, mtmd_gen_inp const*, mtmd_gen_out*) ()
#15 0x0000aaaabee8c88c in mtmd_gen_audio_process ()
#16 0x0000aaaabef10ac8 in qwen3tts_gen_audio_pipeline::step_gen(int, float const*, float const**, bool*) ()
#17 0x0000aaaabef0e438 in mtmd_helper_gen_audio_step_gen ()
#18 0x0000aaaabea7ea48 in main ()
[Inferior 1 (process 761436) detached]
Aborted

**Comment by ServeurpersoCom** [comment]
> > Hello,
> > @Rye426, this should be fixed here: #28231. Note that the mmproj GGUF has to be regenerated, the F32 rule lives in the conversion script.
> 
> Thank you! But There is another issue: when using the Vulkan backend to execute the `GET_ROWS` operator, the backend does not support the current tensor offset configuration. llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:12111: GGML_ASSERT(dst->op != GGML_OP_GET_ROWS || (a_offset == 0 && b_offset == 0 && d_offset == 0)) failed [New LWP ...

Thanks, that is a separate problem: the gen graph hands GET_ROWS an index tensor that is a view at a small byte offset, and the Vulkan path requires those offsets to be aligned, while CPU and CUDA do not care. It hits the 1.7B the same way. 

Please try this fix: https://github.com/ggml-org/llama.cpp/pull/28240

**Comment by Rye426** [comment]
> > > Hello,
> > > @Rye426, this should be fixed here: #28231. Note that the mmproj GGUF has to be regenerated, the F32 rule lives in the conversion script.
> > 
> > 
> > Thank you! But There is another issue: when using the Vulkan backend to execute the `GET_ROWS` operator, the backend does not support the current tensor offset configuration. llama.cpp/ggml/src/ggml-vulkan/ggml-vulkan.cpp:12111: GGML_ASSERT(dst->op != GGML_OP_GET_ROWS || (a_offset == 0 && b_offset == 0 && d_offset == 0)) failed [New LWP ...
> 
> Thanks, that is a separate problem: the gen graph hands GET_ROWS an index tensor that is a view at a small byte offset, and the Vulkan path requires those offsets to be aligned, while CPU and CUDA do not care. It hits the 1.7B the same way.
> 
> Please try this fix: #28240

Thanks
